File: algorithms/0807.max-increase-to-keep-city-skyline/solution.go
func maxIncreaseKeepingSkyline(grid [][]int) int {
	length := len(grid)
	row := make(map[int]int)
	col := make(map[int]int)

	for i := 0; i < length; i++ {
		for j := 0; j < length; j++ {
			val := grid[i][j]

			if val > row[i] {
				row[i] = val
			}
			if val > col[j] {
				col[j] = val
			}
		}
	}

	res := 0
	for i := 0; i < length; i++ {
		for j := 0; j < length; j++ {
			if col[j] > row[i] {
				res += row[i] - grid[i][j]
			} else {
				res += col[j] - grid[i][j]
			}
		}
	}

	return res
}

File: algorithms/0807.max-increase-to-keep-city-skyline/solution.rb
# @param {Integer[][]} grid
# @return {Integer}
def max_increase_keeping_skyline(grid)
  row, col = grid.map(&:max), grid.transpose.map(&:max)

  res = 0
  (0...grid.size).map do |i|
    (0...grid[0].size).map do |j|
      res += [row[i], col[j]].min - grid[i][j]
    end
  end

  res
end
